libs/gltfio/src/MaterialGenerator.h

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright (C) 2019 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef GLTFIO_MATERIALGENERATOR_H
+#define GLTFIO_MATERIALGENERATOR_H
+
+#include <filament/Engine.h>
+#include <filament/Material.h>
+
+#include <utils/Hash.h>
+
+#include <tsl/robin_map.h>
+
+#include <array>
+
+namespace gltfio {
+namespace details {
+
+enum class AlphaMode : uint8_t {
+    OPAQUE,
+    MASKED,
+    TRANSPARENT
+};
+
+struct MaterialKey {
+    bool doubleSided : 1;
+    bool unlit : 1;
+    bool hasVertexColors : 1;
+    bool hasBaseColorTexture : 1;
+    bool hasMetallicRoughnessTexture : 1;
+    bool hasNormalTexture : 1;
+    bool hasOcclusionTexture : 1;
+    bool hasEmissiveTexture : 1;
+    AlphaMode alphaMode;
+    uint8_t baseColorUV;
+    uint8_t metallicRoughnessUV;
+    uint8_t emissiveUV;
+    uint8_t aoUV;
+    uint8_t normalUV;
+    float alphaMaskThreshold;
+    bool hasTextureTransforms;
```

Comment:
I've seen issues before with using `bool`s inside of a `struct` used as a hash key. 

The compiler will align `hasTextureTransforms` to a 4-byte boundary, inserting 3 bytes of padding before it
```
12 34 56 78         // alphaMaskThreshold
XX XX XX 00         // hasTextureTransforms
```

The C++ spec is hairy and haven't confirmed 100%, but I don't think these padding bytes are required to be initialized when using designated initializers:

```
MaterialKey matkey {
    .hasTextureTransforms = true
};
```

This would cause `MurmurHashFn` (which just hashes the bytes) to hash two logically equal `MaterialKey`s differently. I recommend specifying padding explicitly, or doing a `memset` whenever using `MaterialKey`.

Reply:
You can fix this by surrounding the struct with `#pragma pack(push, 1)` and `#pragma pack(pop)`. We've used this for the sort key in the renderer in the past (before we switch to something else for branchless code).

Reply:
Excellent catch, I've run into this issue too.  Should be fixed now.

Reply:
I think you still need to specify the padding at the end of the struct

```
bool hasTextureTransforms;
uint8_t padding[3] = { 0 };
```